File: apps/vitrallis-media-carousel/storage.py
"""Current XDG storage assumption, strict POSIX paths and durable JSON writes."""
import fcntl
import json
import os
from pathlib import Path
import stat
import tempfile

APP_ID = "io.vitrallis.mediacarousel"


def safe_directory(path, private=False):
    raw = os.fspath(path)
    if not raw.startswith("/") or ".." in raw.split("/") or "\x00" in raw:
        raise ValueError("Storage paths must be absolute and contain no traversal")
    path = Path(raw)
    current = Path(path.anchor)
    for part in path.parts[1:]:
        current /= part
        info = current.lstat()
        if not stat.S_ISDIR(info.st_mode):
            raise ValueError("Storage path contains a link or non-directory")
    if private:
        info = path.lstat()
        if info.st_uid != os.getuid() or info.st_mode & 0o077:
            raise ValueError("App storage must be owned by this user with mode 0700")
    return path


def make_private(path):
    path = Path(path)
    if not path.exists() and not path.is_symlink():
        make_private(path.parent)
        path.mkdir(mode=0o700)
    return safe_directory(path)


def regular_open(path, limit):
    path = Path(path)
    safe_directory(path.parent, private=True)
    fd = os.open(path, os.O_RDONLY | os.O_NOFOLLOW | os.O_NONBLOCK)
    try:
        info = os.fstat(fd)
        if (not stat.S_ISREG(info.st_mode) or info.st_nlink != 1
                or info.st_uid != os.getuid() or info.st_size > limit):
            raise ValueError("Unsafe or oversized data file")
        return os.fdopen(fd, "rb")
    except BaseException:
        os.close(fd)
        raise


def unique_keys(pairs):
    result = {}
    for key, value in pairs:
        if key in result:
            raise ValueError("Duplicate JSON key")
        result[key] = value
    return result


def read_json(path, limit):
    with regular_open(path, limit) as stream:
        raw = stream.read(limit + 1)
    if len(raw) > limit:
        raise ValueError("Metadata too large")
    return json.loads(raw, object_pairs_hook=unique_keys,
                      parse_constant=lambda _: (_ for _ in ()).throw(ValueError("Invalid number")))


def sync_directory(path):
    fd = os.open(path, os.O_RDONLY | os.O_DIRECTORY | os.O_NOFOLLOW)
    try:
        os.fsync(fd)
    finally:
        os.close(fd)


def atomic_json(path, data, limit):
    path = Path(path)
    safe_directory(path.parent, private=True)
    payload = (json.dumps(data, ensure_ascii=False, allow_nan=False) + "\n").encode("utf-8")
    if len(payload) > limit:
        raise ValueError("Metadata capacity reached")
    if path.exists() or path.is_symlink():
        with regular_open(path, limit):
            pass
    fd, temporary = tempfile.mkstemp(prefix=".write-", dir=path.parent)
    try:
        with os.fdopen(fd, "wb") as stream:
            stream.write(payload)
            stream.flush()
            os.fsync(stream.fileno())
        os.replace(temporary, path)
        try:
            sync_directory(path.parent)
        except OSError:
            # Rename has committed already. Callers must keep memory aligned with
            # that committed document and surface the durability warning.
            return "Saved, but directory sync failed; durability after power loss is uncertain."
        return ""
    finally:
        if os.path.lexists(temporary):
            os.unlink(temporary)


class Paths:
    def __init__(self, environ=None):
        env = os.environ if environ is None else environ
        home = env.get("HOME", str(Path.home()))
        safe_directory(home)
        self.config = self._root(env.get("XDG_CONFIG_HOME", home + "/.config"))
        self.data = self._root(env.get("XDG_DATA_HOME", home + "/.local/share"))
        self.cache = self._root(env.get("XDG_CACHE_HOME", home + "/.cache"))
        self.media = self.data / "media"
        self.uploads = self.data / "uploads"
        directories = (self.config, self.data, self.cache, self.media, self.uploads)
        # Validate every configured path before creating any directory.
        for path in directories:
            ancestor = path
            while not ancestor.exists() and not ancestor.is_symlink():
                ancestor = ancestor.parent
            safe_directory(ancestor, private=ancestor == path)
        for path in directories:
            make_private(path)
            safe_directory(path, private=True)

    @staticmethod
    def _root(base):
        if not base.startswith("/") or ".." in base.split("/"):
            raise ValueError("Invalid XDG storage location")
        path = Path(base) / APP_ID
        package = Path(__file__).resolve().parent
        if path == package or package in path.parents:
            raise ValueError("Storage must be outside the installed app package")
        return path


class InstanceLock:
    """One metadata owner; never truncate a lock another instance may hold."""
    def __init__(self, directory):
        safe_directory(directory, private=True)
        self.fd = os.open(Path(directory) / "instance.lock",
                          os.O_CREAT | os.O_RDWR | os.O_NOFOLLOW | os.O_NONBLOCK, 0o600)
        try:
            info = os.fstat(self.fd)
            if not stat.S_ISREG(info.st_mode) or info.st_nlink != 1 or info.st_uid != os.getuid():
                raise ValueError("Unsafe instance lock")
            fcntl.flock(self.fd, fcntl.LOCK_EX | fcntl.LOCK_NB)
        except BaseException:
            os.close(self.fd)
            self.fd = None
            raise

    def close(self):
        if self.fd is not None:
            os.close(self.fd)
            self.fd = None

File: apps/carousel-rust/src/storage.rs
//! Shared Python-compatible POSIX storage and durable replacement.
use anyhow::{Context, Result, bail, ensure};
use std::fs::{self, DirBuilder, File, OpenOptions};
use std::io::{Read, Write};
use std::os::unix::fs::{DirBuilderExt, MetadataExt, OpenOptionsExt};
use std::path::{Component, Path, PathBuf};

pub const SHARED_ID: &str = "io.vitrallis.mediacarousel";
pub const MAX_UPLOAD: u64 = 64 * 1024 * 1024;

pub fn uid() -> u32 {
    // SAFETY: getuid has no arguments or memory preconditions.
    unsafe { libc::getuid() }
}

pub fn directory(path: &Path, private: bool) -> Result<()> {
    ensure!(path.is_absolute(), "Storage paths must be absolute");
    let mut current = PathBuf::new();
    for part in path.components() {
        ensure!(!matches!(part, Component::ParentDir), "Storage traversal");
        current.push(part);
        let info = fs::symlink_metadata(&current)?;
        ensure!(info.is_dir(), "Storage contains a link or non-directory");
    }
    if private {
        let info = fs::symlink_metadata(path)?;
        ensure!(
            info.uid() == uid() && info.mode().trailing_zeros() >= 6,
            "Storage must be owned by this user and mode 0700"
        );
    }
    Ok(())
}

fn check_root(path: &Path, package: &Path) -> Result<()> {
    ensure!(
        path.is_absolute() && !path.starts_with(package),
        "Storage must be absolute and outside the package"
    );
    ensure!(
        !path.components().any(|c| matches!(c, Component::ParentDir)),
        "Storage traversal"
    );
    let mut ancestor = path;
    loop {
        match fs::symlink_metadata(ancestor) {
            Ok(_) => return directory(ancestor, ancestor == path),
            Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
                ancestor = ancestor.parent().context("Missing storage ancestor")?;
            }
            Err(e) => return Err(e.into()),
        }
    }
}

fn make_private(path: &Path) -> Result<()> {
    match fs::symlink_metadata(path) {
        Ok(_) => directory(path, false),
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => {
            make_private(path.parent().context("Missing parent")?)?;
            DirBuilder::new().mode(0o700).create(path)?;
            directory(path, false)
        }
        Err(e) => Err(e.into()),
    }
}

#[derive(Clone, Debug)]
pub struct Paths {
    pub config: PathBuf,
    pub data: PathBuf,
    pub media: PathBuf,
    pub uploads: PathBuf,
}

impl Paths {
    pub fn from_env(package: &Path) -> Result<Self> {
        let home = PathBuf::from(std::env::var_os("HOME").context("HOME is required")?);
        directory(&home, false)?;
        let root = |key, fallback: &str| {
            std::env::var_os(key)
                .map_or_else(|| home.join(fallback), PathBuf::from)
                .join(SHARED_ID)
        };
        Self::create(
            root("XDG_CONFIG_HOME", ".config"),
            root("XDG_DATA_HOME", ".local/share"),
            &root("XDG_CACHE_HOME", ".cache"),
            package,
        )
    }

    pub fn create(config: PathBuf, data: PathBuf, cache: &Path, package: &Path) -> Result<Self> {
        let paths = Self {
            media: data.join("media"),
            uploads: data.join("uploads"),
            config,
            data,
        };
        let directories = [
            &paths.config,
            &paths.data,
            cache,
            &paths.media,
            &paths.uploads,
        ];
        // Validate ALL configured roots before the first mutation.
        for path in directories {
            check_root(path, package)?;
        }
        for path in directories {
            make_private(path)?;
            directory(path, true)?;
        }
        Ok(paths)
    }

    pub fn lock(&self) -> Result<File> {
        directory(&self.data, true)?;
        let file = OpenOptions::new()
            .read(true)
            .write(true)
            .create(true)
            .truncate(false)
            .mode(0o600)
            .custom_flags(libc::O_NOFOLLOW | libc::O_NONBLOCK)
            .open(self.data.join("instance.lock"))?;
        check_file(&file, 4096)?;
        // File::try_lock uses flock on Unix, matching Python's fcntl.flock.
        file.try_lock()
            .context("Close the Python or Rust carousel before opening another instance")?;
        Ok(file)
    }
}

fn check_file(file: &File, limit: u64) -> Result<()> {
    let info = file.metadata()?;
    ensure!(
        info.is_file() && info.nlink() == 1 && info.uid() == uid() && info.len() <= limit,
        "Unsafe or oversized data file"
    );
    Ok(())
}

pub fn regular(path: &Path, limit: u64) -> Result<File> {
    directory(path.parent().context("Missing parent")?, true)?;
    let file = OpenOptions::new()
        .read(true)
        .custom_flags(libc::O_NOFOLLOW | libc::O_NONBLOCK)
        .open(path)?;
    check_file(&file, limit)?;
    Ok(file)
}

pub fn read(path: &Path, limit: u64) -> Result<Vec<u8>> {
    let mut bytes = Vec::new();
    regular(path, limit)?
        .take(limit + 1)
        .read_to_end(&mut bytes)?;
    ensure!(
        u64::try_from(bytes.len())? <= limit,
        "File grew beyond limit"
    );
    Ok(bytes)
}

pub fn random_id() -> String {
    format!("{:032x}", rand::random::<u128>())
}

pub struct Temporary {
    pub path: PathBuf,
    pub file: File,
}

impl Temporary {
    pub fn new(parent: &Path, prefix: &str) -> Result<Self> {
        directory(parent, true)?;
        for _ in 0..8 {
            let path = parent.join(format!("{prefix}{}", random_id()));
            match OpenOptions::new()
                .read(true)
                .write(true)
                .create_new(true)
                .mode(0o600)
                .open(&path)
            {
                Ok(file) => return Ok(Self { path, file }),
                Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => (),
                Err(e) => return Err(e.into()),
            }
        }
        bail!("Temporary file collision")
    }
}

impl Drop for Temporary {
    fn drop(&mut self) {
        if let Err(e) = fs::remove_file(&self.path)
            && e.kind() != std::io::ErrorKind::NotFound
        {
            eprintln!("event=cleanup_failed error={e}");
        }
    }
}

pub fn sync(path: &Path) -> Result<()> {
    directory(path, true)?;
    File::open(path)?.sync_all()?;
    Ok(())
}

pub fn atomic_json<T: serde::Serialize>(path: &Path, value: &T, limit: u64) -> Result<String> {
    let mut bytes = serde_json::to_vec(value)?;
    bytes.push(b'\n');
    ensure!(
        u64::try_from(bytes.len())? <= limit,
        "Metadata capacity reached"
    );
    match fs::symlink_metadata(path) {
        Ok(_) => {
            regular(path, limit)?;
        }
        Err(e) if e.kind() == std::io::ErrorKind::NotFound => (),
        Err(e) => return Err(e.into()),
    }
    let parent = path.parent().context("Missing parent")?;
    let mut temp = Temporary::new(parent, ".write-")?;
    temp.file.write_all(&bytes)?;
    temp.file.sync_all()?;
    fs::rename(&temp.path, path)?;
    // Rename is the commit point. A later fsync failure must not roll back memory.
    Ok(match sync(parent) {
        Ok(()) => String::new(),
        Err(_) => "Saved, but directory sync failed; power-loss durability is uncertain.".into(),
    })
}

/// Atomically publish a staged file without replacing a concurrent destination.
pub fn publish(source: &Path, destination: &Path) -> Result<()> {
    use std::os::unix::ffi::OsStrExt;
    directory(source.parent().context("Missing source parent")?, true)?;
    directory(
        destination.parent().context("Missing destination parent")?,
        true,
    )?;
    let source = std::ffi::CString::new(source.as_os_str().as_bytes())?;
    let destination = std::ffi::CString::new(destination.as_os_str().as_bytes())?;
    #[cfg(target_os = "linux")]
    // SAFETY: valid NUL-terminated paths; no-replace rename has no pointer writes.
    let result = unsafe {
        libc::renameat2(
            libc::AT_FDCWD,
            source.as_ptr(),
            libc::AT_FDCWD,
            destination.as_ptr(),
            libc::RENAME_NOREPLACE,
        )
    };
    #[cfg(target_os = "macos")]
    // SAFETY: valid NUL-terminated paths; RENAME_EXCL forbids replacement.
    let result =
        unsafe { libc::renamex_np(source.as_ptr(), destination.as_ptr(), libc::RENAME_EXCL) };
    #[cfg(not(any(target_os = "linux", target_os = "macos")))]
    let result = -1;
    ensure!(
        result == 0,
        "Cannot publish staged file: {}",
        std::io::Error::last_os_error()
    );
    Ok(())
}
